rust/kernel/proc_fs.rs

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Type for defining `proc` files.
+//!
+//! This module allows Rust devices to create entries in `/proc` from a
+//! [`bindings::proc_ops`] vtable.
+//!
+//! C header: [`include/linux/proc_fs.h`](../../../include/linux/proc_fs.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/proc.html>
+
+use core::{
+    marker::{PhantomData, Sync},
+    ops::Deref,
+    ptr,
+};
+
+use crate::{bindings, c_types, types::PointerWrapper, CStr, Error, Result};
+
+/// An entry under `/proc` containing data of type `T`.
+///
+/// This is the Rust equivalent to [`proc_dir_entry`] on the C side.
+///
+/// # Invariants
+///
+/// The [`ProcDirEntry::proc_dir_entry`] is a valid pointer.
+/// [`ProcDirEntry::data`] points to the PDE data of
+/// [`ProcDirEntry::proc_dir_entry`].
+/// [`ProcDirEntry::data`] was created by a call to `T::into_pointer`.
+///
+/// [`proc_dir_entry`]: ../../../fs/proc/internal.h
+pub struct ProcDirEntry<T: PointerWrapper> {
+    proc_dir_entry: *mut bindings::proc_dir_entry,
+    data: *const c_types::c_void,
```

Comment:
It's a shame that we don't have access to `proc_dir_entry->data`... (No action needed on your side, I'm just bemoaning about the interface exposed by the C side.)

Reply:
Yeah, I was a bit surprised too...